packages/next/src/client/components/router-reducer/reducers/server-action-reducer.ts

```diff
@@ -131,113 +135,125 @@ export function serverActionReducer(
   state: ReadonlyReducerState,
   action: ServerActionAction
 ): ReducerState {
-  // the action could be called twice so we need to check if we already have applied it
-  if (action.mutable.serverActionApplied) {
-    return state
+  const { mutable, cache, resolve } = action
+  const href = state.canonicalUrl
+
+  let currentTree = state.tree
+
+  const isForCurrentTree =
+    JSON.stringify(mutable.previousTree) === JSON.stringify(currentTree)
+
+  if (isForCurrentTree) {
+    return handleMutable(state, mutable)
   }
 
   if (!action.mutable.inFlightServerAction) {
-    action.mutable.previousTree = state.tree
-    action.mutable.previousUrl = state.canonicalUrl
     action.mutable.inFlightServerAction = createRecordFromThenable(
       fetchServerAction(state, action)
     )
   }
-  try {
-    // suspends until the server action is resolved.
-    const {
-      actionResult,
-      actionFlightData,
-      redirectLocation,
-      revalidatedParts,
-    } = readRecordValue(
-      action.mutable.inFlightServerAction!
-    ) as Awaited<FetchServerActionResult>
-
-    // Invalidate the cache for the revalidated parts. This has to be done before the
-    // cache is updated with the action's flight data again.
-    if (revalidatedParts.tag || revalidatedParts.cookie) {
-      // Invalidate everything if the tag is set.
-      state.prefetchCache.clear()
-    } else if (revalidatedParts.paths.length > 0) {
-      // Invalidate all subtrees that are below the revalidated paths, and invalidate
-      // all the prefetch cache.
-      // TODO-APP: Currently the prefetch cache doesn't have subtree information,
-      // so we need to invalidate the entire cache if a path was revalidated.
-      state.prefetchCache.clear()
+
+  // suspends until the server action is resolved.
+  const {
+    actionResult,
+    actionFlightData: flightData,
+    redirectLocation,
+    // revalidatedParts,
+  } = readRecordValue(
+    action.mutable.inFlightServerAction!
+  ) as Awaited<FetchServerActionResult>
+
+  mutable.previousTree = state.tree
+
+  if (!flightData) {
+    if (!mutable.actionResultResolved) {
+      resolve(actionResult)
+      mutable.actionResultResolved = true
     }
 
+    // If there is a redirect but no flight data we need to do a mpaNavigation.
     if (redirectLocation) {
-      // the redirection might have a flight data associated with it, so we'll populate the cache with it
-      if (actionFlightData) {
-        const href = createHrefFromUrl(redirectLocation, false)
-        const previousCacheEntry = state.prefetchCache.get(href)
-        state.prefetchCache.set(href, {
```

Comment:
I think you need to keep this logic

Reply:
Since it's applied immediately this case is more similar to loading the page entirely fresh, the prefetch cache doesn't have to be seeded for it I think.